### DISCUSSION TITLE
Where is the doc for https://github.com/google/or-tools/blob/stable/examples/python/qubo_sat.py ?

### DISCUSSION THREAD

**Discussion by keithbriggs** [google/or-tools#3127] [CP-SAT questions] [answered]
The example at https://github.com/google/or-tools/blob/stable/examples/python/qubo_sat.py is very interesting.   However, I can't find any documentation on it.   What exactly does the matrix RAW_DATA represent?    How does the solution method work?  Any pointers to answers to these questions would be appreciated.

**Comment by lperron** [comment]
this is a qubo problem: https://en.wikipedia.org/wiki/Quadratic_unconstrained_binary_optimization

the matrix is the closeness of two nodes.
The objective is to minimize sum(x[i]*x[j]*raw_data[i, j]).
The solution tries to solve it with CP-SAT.

**Comment by keithbriggs** [comment]
Thank you, but I don't think it can be a simple as that.  For one thing, RAW_DATA is not a square matrix.   It has one more column than rows.  I think the right-most column is treated as the coefficients of a linear term in the objective (which you don't write in your answer).
Secondly, what is CP-SAT solving?   Is it solving a MAX-SAT problem, or just a normal SAT with some kind of outer loop with maybe a bisection method?   I understand QUBO on its own, and I understand SAT and MAX-SAT, but I am unclear on the relation between these problems.
(Also, there is a right bracket missing in your answer.)

**Comment by lperron** [comment]
true, there is an extra cost for each variables at then end of each line of the matrix, so
```
sum(x[i]*x[j]*raw_data[i, j]) + sum(x[i] * raw_data[i][-1])
```

CP-SAT is a implementation of a constraint programming/MIP solver on top of a SAT backend.

We gave a talk on youtube at CP-AI-OR 2020 masterclass: https://www.youtube.com/watch?v=lmy1ddn4cyw

**Comment by keithbriggs** [comment]
Ok, but could you document what range of values i and j cover in that sum?  Also, in the double loop:

```
    for i in range(num_vars-1):
        x_i = variables[i]
        for j in range(i + 1, num_vars):
           ...
```

we never get i==j, so it looks like the squared terms are omitted from the objective (we get only the cross terms).  Is this as intended?  Also, should the j loop maybe be` for j in range(i+1, num_vars-1):`?

**Reply by lperron** [reply]
I cover the diagonal in the second loop.

As x[i] * x[j] == x[j] * x[i], I only scan one half of the matrix, and add x[i]*x[j] * (raw_data[i][j] * raw_data[j][i])

**Comment by keithbriggs** [comment]
Do you mean `x[i]*x[j] * (raw_data[i][j] + raw_data[j][i])`?

I think it would be a much clearer example if there was a comment block stating **precisely** what mathematical problem was being solved.  

And maybe even clearer if a numpy square matrix `A` and a separate numpy vector `b` were used for the coefficients.  Then if we just said that we were minimizing `x.T@A@x+b@x` (with `x` a 0-1 vector), it would be absolutely clear and precise (even if a slightly different formulation).

Also, it would be good to know the right answer for the RAW_DATA example, so we know if it is working correctly.

**Comment by lperron** [comment, marked as answer]
indeed, that would be nice.

A bit hard to do in addition to my full time job (which is not or-tools), and the flow of users requests from or-tools (which I answer mostly on my family time).

**Comment by keithbriggs** [comment]
Ok, thank you.  I will do it myself.

**Comment by lperron** [comment]
PR are welcome :-)

**Comment by keithbriggs** [comment]
I can now answer my own question at the start of this thread.   There **is** documentation on this method - it is at https://developers.google.com/optimization/cp/cp_solver.   It works by finding all feasible solutions and checking the objective at each.
